r/src/arrow_cpp11.h

```diff
@@ -300,22 +297,65 @@ bool GetBoolOption(const std::string& name, bool default_);
 namespace cpp11 {
 
 template <typename T>
-using enable_if_shared_ptr = typename std::enable_if<
-    std::is_same<std::shared_ptr<typename T::element_type>, T>::value, T>::type;
+std::string r6_class_name(const std::shared_ptr<T>& x);
```

Comment:
Based on the way this is specialized below, it seems we should prefer a trait struct instead of a function to make future extension easier.
```suggestion
struct r6_class_name;
```

Specialized like:
```c++
template <>
struct r6_class_name<arrow::Field> {
  static const char* get(const Field&) { return "Field"; }
};
// ...
template <>
struct r6_class_name<arrow::Array> {
  static const char* get(const Array& array) {
    switch (array.type_id()) {
      case arrow::Type::DICTIONARY:
        return "DictionaryArray";
    //...
    }
  }
};
```

Reply:
Is there any way this could get a default using `nameof()`? I think that'd cover most of the cases like `Field` which don't require dispatch

Reply:
Oh interesting, I had not seen `nameof()` before.